File: android-app/app/src/main/java/eu/kanade/tachiyomi/network/interceptor/SpecificHostRateLimitInterceptor.java
package eu.kanade.tachiyomi.network.interceptor;

import java.util.concurrent.TimeUnit;
import kotlin.time.Duration;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;

/**
 * Compatibility class for older extensions that expect
 * eu.kanade.tachiyomi.network.interceptor.SpecificHostRateLimitInterceptor
 * to contain static rateLimitHost methods (instead of SpecificHostRateLimitInterceptorKt).
 *
 * This class provides all known signatures, including reverse-order (HttpUrl, Builder)
 * that some extensions compiled against.
 */
public final class SpecificHostRateLimitInterceptor {
    private SpecificHostRateLimitInterceptor() {}

    // Builder first, HttpUrl second (new API)
    public static OkHttpClient.Builder rateLimitHost(OkHttpClient.Builder builder, HttpUrl httpUrl, int permits, long period, TimeUnit unit) {
        return builder.addInterceptor(new RateLimitInterceptor(httpUrl.host(), permits, period, unit));
    }

    public static OkHttpClient.Builder rateLimitHost(OkHttpClient.Builder builder, HttpUrl httpUrl, int permits, Duration period) {
        // Duration is inline value class - try to extract millis via getInWholeMilliseconds()
        long millis = 1000L;
        try {
            millis = period.getInWholeMilliseconds();
        } catch (Throwable ignored) {
            try {
                // fallback via toString parsing or default
                millis = 1000L;
            } catch (Throwable ignored2) {}
        }
        return builder.addInterceptor(new RateLimitInterceptor(httpUrl.host(), permits, millis, TimeUnit.MILLISECONDS));
    }

    public static OkHttpClient.Builder rateLimitHost(OkHttpClient.Builder builder, String url, int permits, long period, TimeUnit unit) {
        String host = url;
        try {
            HttpUrl parsed = HttpUrl.parse(url);
            if (parsed != null) host = parsed.host();
        } catch (Exception ignored) {}
        return builder.addInterceptor(new RateLimitInterceptor(host, permits, period, unit));
    }

    public static OkHttpClient.Builder rateLimitHost(OkHttpClient.Builder builder, String url, int permits, Duration period) {
        String host = url;
        try {
            HttpUrl parsed = HttpUrl.parse(url);
            if (parsed != null) host = parsed.host();
        } catch (Exception ignored) {}
        long millis = 1000L;
        try {
            millis = period.getInWholeMilliseconds();
        } catch (Throwable ignored) {}
        return builder.addInterceptor(new RateLimitInterceptor(host, permits, millis, TimeUnit.MILLISECONDS));
    }

    // Reverse order: HttpUrl first, Builder second (old extensions)
    public static OkHttpClient.Builder rateLimitHost(HttpUrl httpUrl, OkHttpClient.Builder builder, int permits, long period, TimeUnit unit) {
        return rateLimitHost(builder, httpUrl, permits, period, unit);
    }

    public static OkHttpClient.Builder rateLimitHost(HttpUrl httpUrl, OkHttpClient.Builder builder, int permits, Duration period) {
        return rateLimitHost(builder, httpUrl, permits, period);
    }

    public static OkHttpClient.Builder rateLimitHost(String url, OkHttpClient.Builder builder, int permits, long period, TimeUnit unit) {
        return rateLimitHost(builder, url, permits, period, unit);
    }

    public static OkHttpClient.Builder rateLimitHost(String url, OkHttpClient.Builder builder, int permits, Duration period) {
        return rateLimitHost(builder, url, permits, period);
    }

    // Factory versions returning interceptor (some old code called these)
    public static RateLimitInterceptor rateLimitHost(HttpUrl httpUrl, int permits, long period, TimeUnit unit) {
        return new RateLimitInterceptor(httpUrl.host(), permits, period, unit);
    }

    public static RateLimitInterceptor rateLimitHost(HttpUrl httpUrl, int permits, Duration period) {
        long millis = 1000L;
        try { millis = period.getInWholeMilliseconds(); } catch (Throwable ignored) {}
        return new RateLimitInterceptor(httpUrl.host(), permits, millis, TimeUnit.MILLISECONDS);
    }

    public static RateLimitInterceptor rateLimitHost(String host, int permits, long period, TimeUnit unit) {
        return new RateLimitInterceptor(host, permits, period, unit);
    }

    public static RateLimitInterceptor rateLimitHost(String host, int permits, Duration period) {
        long millis = 1000L;
        try { millis = period.getInWholeMilliseconds(); } catch (Throwable ignored) {}
        return new RateLimitInterceptor(host, permits, millis, TimeUnit.MILLISECONDS);
    }
}

File: android-app/app/src/main/kotlin/eu/kanade/tachiyomi/network/interceptor/SpecificHostRateLimitInterceptor.kt
@file:JvmName("SpecificHostRateLimitInterceptor")
package eu.kanade.tachiyomi.network.interceptor

import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import java.util.concurrent.TimeUnit
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds
import kotlin.time.toDuration
import kotlin.time.toDurationUnit

// New API (Builder receiver) -> generates static methods in class SpecificHostRateLimitInterceptor
fun OkHttpClient.Builder.rateLimitHost(
    httpUrl: HttpUrl,
    permits: Int,
    period: Long = 1,
    unit: TimeUnit = TimeUnit.SECONDS,
): OkHttpClient.Builder = addInterceptor(RateLimitInterceptor(httpUrl.host, permits, period.toDuration(unit.toDurationUnit())))

fun OkHttpClient.Builder.rateLimitHost(
    httpUrl: HttpUrl,
    permits: Int,
    period: Duration = 1.seconds,
): OkHttpClient.Builder = addInterceptor(RateLimitInterceptor(httpUrl.host, permits, period))

fun OkHttpClient.Builder.rateLimitHost(
    url: String,
    permits: Int,
    period: Long = 1,
    unit: TimeUnit = TimeUnit.SECONDS,
): OkHttpClient.Builder {
    val host = try { HttpUrl.get(url).host } catch (_: Exception) { url }
    return addInterceptor(RateLimitInterceptor(host, permits, period.toDuration(unit.toDurationUnit())))
}

fun OkHttpClient.Builder.rateLimitHost(
    url: String,
    permits: Int,
    period: Duration = 1.seconds,
): OkHttpClient.Builder {
    val host = try { HttpUrl.get(url).host } catch (_: Exception) { url }
    return addInterceptor(RateLimitInterceptor(host, permits, period))
}

// Reverse order overloads for old extensions that expect (HttpUrl, Builder) order
fun HttpUrl.rateLimitHost(
    builder: OkHttpClient.Builder,
    permits: Int,
    period: Long = 1,
    unit: TimeUnit = TimeUnit.SECONDS,
): OkHttpClient.Builder = builder.addInterceptor(RateLimitInterceptor(host, permits, period.toDuration(unit.toDurationUnit())))

fun HttpUrl.rateLimitHost(
    builder: OkHttpClient.Builder,
    permits: Int,
    period: Duration = 1.seconds,
): OkHttpClient.Builder = builder.addInterceptor(RateLimitInterceptor(host, permits, period))

fun String.rateLimitHost(
    builder: OkHttpClient.Builder,
    permits: Int,
    period: Long = 1,
    unit: TimeUnit = TimeUnit.SECONDS,
): OkHttpClient.Builder {
    val h = try { HttpUrl.get(this).host } catch (_: Exception) { this }
    return builder.addInterceptor(RateLimitInterceptor(h, permits, period.toDuration(unit.toDurationUnit())))
}

fun String.rateLimitHost(
    builder: OkHttpClient.Builder,
    permits: Int,
    period: Duration = 1.seconds,
): OkHttpClient.Builder {
    val h = try { HttpUrl.get(this).host } catch (_: Exception) { this }
    return builder.addInterceptor(RateLimitInterceptor(h, permits, period))
}

// Provide class-based shims for extensions that call SpecificHostRateLimitInterceptor.rateLimitHost as static factory returning interceptor
object SpecificHostRateLimitInterceptorCompat {
    @JvmStatic
    fun rateLimitHost(host: String, permits: Int, period: Long, unit: TimeUnit): RateLimitInterceptor {
        return RateLimitInterceptor(host, permits, period.toDuration(unit.toDurationUnit()))
    }
    @JvmStatic
    fun rateLimitHost(url: HttpUrl, permits: Int, period: Duration): RateLimitInterceptor {
        return RateLimitInterceptor(url.host, permits, period)
    }
    @JvmStatic
    fun rateLimitHost(httpUrl: HttpUrl, permits: Int, period: Long, unit: TimeUnit): RateLimitInterceptor {
        return RateLimitInterceptor(httpUrl.host, permits, period.toDuration(unit.toDurationUnit()))
    }
    @JvmStatic
    fun rateLimitHost(host: String, permits: Int, period: Duration): RateLimitInterceptor {
        return RateLimitInterceptor(host, permits, period)
    }
}
